### PR TITLE
Long GrainReferences break AzureTableStorage Provider

### PR DESCRIPTION
**Pull request by zcarson**
Two new unit tests to show that grains with long grainreference values will fail to write to AzureTable. The issue appears at or about 190 characters. This seems to happen most often in grains that use generics, because the type information is expanded when the grainreference is created. A grain with generics that uses `<int>`, for example, expands to `[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]`, which is an additional 90 characters.

### CODE REVIEW

File: src/Tester/GenericGrainTests.cs [hunk 1]
```diff
@@ -287,6 +293,20 @@ public static void MyClassCleanup()
         }
 
         [TestMethod, TestCategory("BVT"), TestCategory("Functional"), TestCategory("Generics")]
```
Comment (jason-bragg) [inline src/Tester/GenericGrainTests.cs]: Tests with external dependencies, like azure, should not be marked as BVT.

**Comment by jason-bragg** [comment]
@zcarson  Thanks for these tests.
I wouldn't expect this behavior since azure storage partition and row keys can be fairly large (https://msdn.microsoft.com/en-us/library/dd179338.aspx), but something's obviously amiss.

File: src/OrleansTestingHost/OrleansConfigurationForTesting.xml [hunk 2]
```diff
@@ -4,7 +4,7 @@
     <StorageProviders>
       <Provider Type="Orleans.Storage.MemoryStorage" Name="MemoryStore" />
       <Provider Type="Orleans.Storage.MemoryStorage" Name="Default" />
-      <!--<Provider Type="Orleans.Storage.AzureTableStorage" Name="AzureStore" />-->
+      <Provider Type="Orleans.Storage.AzureTableStorage" Name="AzureStore" DataConnectionString="UseDevelopmentStorage=true"/>
```
Comment (gabikliot) [inline src/OrleansTestingHost/OrleansConfigurationForTesting.xml]: please do not change this in this file. This file is used by a lot of other tests. By addding this lien you now caused all other unrelated tests to load AzureStore provider, and potentially try to connect to dev storage.
Instead, for your test only, you can use programmable APIs to tweak configuration.
Look for example into ControllableStreamProviderTests.AdjustForTest.

File: src/Tester/GenericGrainTests.cs (hunk 1 above)
Comment (gabikliot) [inline src/Tester/GenericGrainTests.cs]: Also, please mark the failing test with Ignore attribute. Otherwise, it will fail CI.

**Comment by jason-bragg** [comment]
@zcarson Looks good but needs to be rebased.  Since this needs an update for rebase anyway, can you tag the added tests with TestCategory("Azure") as well.

Thanks!

**Comment by zcarson** [comment]
@jason-bragg will do

**Comment by jason-bragg** [comment]
@zcarson 
Thanks for the tests.

From my investigation, this looks to only be an issue when testing against the storage emulator.  The emulator is not a perfect simulator of the storage service. :/

When testing against azure storage service, I was able to create entities with partition keys as large as 500 characters (500 * 2 (size of Unicode char) = 1kb).

**Comment by gabikliot** [comment]
https://azure.microsoft.com/en-us/documentation/articles/storage-use-emulator/#differences-between-the-storage-emulator-and-azure-storage

"The storage emulator supports partition key and row key property values of less than 512 bytes each. Additionally, the total size of the account name, table name, and key property names together cannot exceed 900 bytes."

While in the service the PK and RK can be up to 1K:
https://msdn.microsoft.com/en-us/library/dd179338.aspx
"The partition key may be a string value up to 1 KB in size."
"The row key is a string value that may be up to 1 KB in size."

**Pull request merged**